### PR TITLE
cli: bump to jest 27

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

This bumps jest to 27, as well as implements configuration detection based on package role according to #8729

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by freben — APPROVED** [review]
Sweet!

**Review by adamdmharvey — APPROVED** [review]
👍🏻 for CircleCI

**Review by camilaibs — APPROVED** [review]

**Comment by Rugvip** [comment]
Dependencies for this has been flipped around, we need to roll out #9254 along with migration to package roles across the board before this can move forward. It's due to the removals of certain APIs in the `jsdom` environment in Jest 27, meaning that once we move to Jest 27 the environment needs to be set to either `jsdom` or `node` depending on the package. Not having it set correctly is likely to lead to breakages on both ends.

We could switch to Jest 27 directly and require that the environment is set in config, but we'd rather roll out package roles and make that the primary option instead, to reduce the number of migrations needed.

**Comment by Rugvip** [comment]
#10064 enables us to ship this. But still want to wait a bit longer to not force people to migrate too soon

**Comment by backstage-service** [comment]
## Missing Changesets

The following package(s) are changed by this PR but do not have a changeset:

- **@backstage/backend-tasks**
- **@backstage/backend-test-utils**
- **@backstage/plugin-tech-radar**

See [CONTRIBUTION.md](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets) for more information about how to add changesets.


## Changed Packages

| Package Name | Package Path | Changeset Bump | Current Version |
|:-------------|:-------------|:--------------:|:----------------|
| @backstage/backend-common | packages/backend-common | **patch** | `v0.13.2-next.0` |
| @backstage/backend-tasks | packages/backend-tasks | **none** | `v0.2.2-next.0` |
| @backstage/backend-test-utils | packages/backend-test-utils | **none** | `v0.1.23-next.0` |
| @backstage/cli | packages/cli | **minor** | `v0.16.1-next.0` |
| @backstage/plugin-tech-radar | plugins/tech-radar | **none** | `v0.5.11-next.0` |

**Comment by backstage-service** [comment]
## Missing Changesets

The following package(s) are changed by this PR but do not have a changeset:

- **@backstage/backend-tasks**
- **@backstage/backend-test-utils**
- **@backstage/plugin-kubernetes-backend**
- **@backstage/plugin-tech-radar**

See [CONTRIBUTION.md](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets) for more information about how to add changesets.


## Changed Packages

| Package Name | Package Path | Changeset Bump | Current Version |
|:-------------|:-------------|:--------------:|:----------------|
| @backstage/backend-common | packages/backend-common | **patch** | `v0.13.2-next.0` |
| @backstage/backend-tasks | packages/backend-tasks | **none** | `v0.2.2-next.0` |
| @backstage/backend-test-utils | packages/backend-test-utils | **none** | `v0.1.23-next.0` |
| @backstage/cli | packages/cli | **minor** | `v0.16.1-next.0` |
| @backstage/plugin-kubernetes-backend | plugins/kubernetes-backend | **none** | `v0.4.14-next.0` |
| @backstage/plugin-tech-radar | plugins/tech-radar | **none** | `v0.5.11-next.0` |

File: packages/backend-tasks/src/tasks/PluginTaskSchedulerImpl.test.ts [hunk 1]
```diff
@@ -48,16 +48,16 @@ describe('PluginTaskManagerImpl', () => {
         const { manager } = await init(databaseId);
 
         const fn = jest.fn();
+        const promise = new Promise(resolve => fn.mockImplementation(resolve));
```
Comment (freben) [inline packages/backend-tasks/src/tasks/PluginTaskSchedulerImpl.test.ts:51]: This is actually pretty neat

File: packages/backend-tasks/src/tasks/TaskWorker.ts [hunk 2]
```diff
@@ -63,7 +59,7 @@ export class TaskWorker {
             break;
           }
 
-          await sleep(WORK_CHECK_FREQUENCY, options?.signal);
+          await sleep(this.workCheckFrequency, options?.signal);
```
Comment (freben) [inline packages/backend-tasks/src/tasks/TaskWorker.ts:62]: You forgot a changeset for this package right?

Comment (Rugvip) [reply packages/backend-tasks/src/tasks/TaskWorker.ts:62]: Yep!

**Pull request merged**